main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -17,15 +17,20 @@ package ca.uwaterloo.flix.language.phase
 
 import ca.uwaterloo.flix.api.Flix
 import ca.uwaterloo.flix.language.CompilationMessage
-import ca.uwaterloo.flix.language.ast.{Ast, ReadAst, Token}
+import ca.uwaterloo.flix.language.ast.{Ast, ErrKind, ReadAst, SourceKind, SourceLocation, Token, TokenKind}
+import ca.uwaterloo.flix.language.errors.LexerError
 import ca.uwaterloo.flix.util.{ParOps, Validation}
 import ca.uwaterloo.flix.util.Validation._
 
 import scala.collection.mutable
 
 object Lexer {
 
-  def run(root: ReadAst.Root)(implicit flix: Flix): Validation[Map[Ast.Source, Array[Token]], CompilationMessage] = {
+  // TODO: A single source can have more than one error that we would like to report.
+  // For instance there might be an unexpected char multiple spots in the source file.
+  // Therefore run needs to return something like `Validation[Map[Ast.Source, Array[Token]], Array[CompilationMessage]]`
+  // but that seems to clash with the rest of the pipeline. How do we resolve this?
+  def run(root: ReadAst.Root)(implicit flix: Flix): Validation[Map[Ast.Source, Array[Token]], Array[CompilationMessage]] = {
```

Comment:
This should not be Array. A validation (e.g. failure and softFailure) already supports multiple failures.

Reply:
I'm having trouble getting this to work. It seems to me like I need to use `Validation.sequence` and maybe `Validation.mapN`?

Reply:
In that case you probably need to construct `SoftFailure` manually.

Reply:
Yep that did the trick! I ended up with 
```scala
    Validation.SoftFailure(s.tokens.toArray, LazyList.from(s.tokens).collect {
      case Token(TokenKind.Err(e), t, l, c) => tokenErrToCompilationMessage(e, t, l, c)
    })
```